packages/commonwealth/client/scripts/state/api/comments/createComment.ts

```diff
@@ -0,0 +1,100 @@
+import { useMutation, useQueryClient } from '@tanstack/react-query';
+import axios from 'axios';
+import { updateLastVisited } from 'controllers/app/login';
+import Comment from 'models/Comment';
+import Thread from 'models/Thread';
+import app from 'state';
+import { ApiEndpoints } from 'state/api/config';
+import useFetchCommentsQuery from './fetchComments';
+
+interface CreateCommentProps {
+  address: string;
+  threadId: number,
+  chainId: string;
+  unescapedText: string,
+  parentCommentId: any,
+  attachments?: string[]
+}
+
+const createComment = async ({
+  chainId,
+  address,
+  threadId,
+  attachments,
+  unescapedText,
+  parentCommentId = null,
+}: CreateCommentProps) => {
+  const {
+    session = null,
+    action = null,
+    hash = null,
+  } = await app.sessions.signComment({
+    thread_id: threadId,
+    body: unescapedText,
+    parent_comment_id: parentCommentId,
+  });
+
+  const response = await axios.post(
+    `${app.serverUrl()}/threads/${threadId}/comments`,
+    {
+      author_chain: chainId,
+      chain: chainId,
+      address: address,
+      parent_id: parentCommentId,
+      // chain_entity_id: chainEntity?.id,
+      'attachments[]': attachments,
+      text: encodeURIComponent(unescapedText),
+      jwt: app.user.jwt,
+      canvas_action: action,
+      canvas_session: session,
+      canvas_hash: hash,
+    }
+  );
+
+  return new Comment(response.data.result)
+};
+
+const useCreateCommentMutation = ({ chainId, threadId }: Partial<CreateCommentProps>) => {
+  const queryClient = useQueryClient();
+  const { data: comments } = useFetchCommentsQuery({
+    chainId,
+    threadId,
+  })
+
+  return useMutation({
+    mutationFn: createComment,
+    onSuccess: async (newComment) => {
+      // update fetch comments query state
+      const key = [ApiEndpoints.FETCH_COMMENTS, chainId, threadId]
+      queryClient.cancelQueries({ queryKey: key });
+      queryClient.setQueryData([...key],
+        () => {
+          const updatedComments = [...(comments || []).filter(x => x.id !== newComment.id), newComment]
```

Comment:
> Is there a situation when the comments will be other than an array

no, it will always be an array (until we add the pagination refactor after its complete, in which case there will be a different format)

> How about this?
return comments.filter(x => x.id !== newComment.id), newComment

Since `comments` will always be an array so we will need to return an array here. 

Another option is to do mutation, like `return comments.filter(x => x.id !== newComment.id).push(newComment)` but I think it's better to avoid mutations since the `comments` is at the end stored in react query cache and the cache might mess up with mutations (obj reference will be the same with mutations)

Reply:
> > Is there a situation when the comments will be other than an array
> 
> 
> no, it will always be an array (until we add the pagination refactor after its complete, in which case there will be a different format)
> 

If this is always array, we do not need to do `comments || []`

> Since comments will always be an array so we will need to return an array here.

wait a second... 
`comments.filter()` actually returns an array with comments. But I start thinking, why you need to filter existing comments when you add a new comment?

Reply:
> wait a second...
comments.filter() actually returns an array with comments. But I start thinking, why you need to filter existing comments when you add a new comment?

ah, I see now, I don't know why I did this but yes filtering is not needed (maybe I just copied over the editComment RQ code or something). Updated

Thanks for the spot.